File: src/engine/pieces.py
"""Data classes for all game pieces on the board."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum, auto


class Terrain(Enum):
    JUNGLE = auto()
    MOUNTAIN = auto()
    SANDS = auto()
    WETLAND = auto()


class InvaderType(Enum):
    EXPLORER = auto()
    TOWN = auto()
    CITY = auto()


# Health and damage values per invader type
INVADER_HEALTH = {
    InvaderType.EXPLORER: 1,
    InvaderType.TOWN: 2,
    InvaderType.CITY: 3,
}

INVADER_DAMAGE = {
    InvaderType.EXPLORER: 1,
    InvaderType.TOWN: 2,
    InvaderType.CITY: 3,
}

# Fear generated when destroying an invader
INVADER_FEAR = {
    InvaderType.EXPLORER: 0,
    InvaderType.TOWN: 1,
    InvaderType.CITY: 2,
}

DAHAN_HEALTH = 2
DAHAN_DAMAGE = 2


@dataclass
class Invader:
    type: InvaderType
    damage_taken: int = 0

    @property
    def health(self) -> int:
        return INVADER_HEALTH[self.type]

    @property
    def remaining_health(self) -> int:
        return self.health - self.damage_taken

    @property
    def is_destroyed(self) -> bool:
        return self.damage_taken >= self.health

    @property
    def damage_output(self) -> int:
        return INVADER_DAMAGE[self.type]

    @property
    def fear_on_destroy(self) -> int:
        return INVADER_FEAR[self.type]


@dataclass
class Dahan:
    damage_taken: int = 0

    @property
    def remaining_health(self) -> int:
        return DAHAN_HEALTH - self.damage_taken

    @property
    def is_destroyed(self) -> bool:
        return self.damage_taken >= DAHAN_HEALTH


class Element(Enum):
    SUN = auto()
    MOON = auto()
    FIRE = auto()
    AIR = auto()
    WATER = auto()
    EARTH = auto()
    PLANT = auto()
    ANIMAL = auto()


class PowerSpeed(Enum):
    FAST = auto()
    SLOW = auto()


@dataclass
class PowerCard:
    name: str
    cost: int
    speed: PowerSpeed
    range: int
    target: str
    elements: list[Element] = field(default_factory=list)
    description: str = ""

File: src/engine/pieces.ts
export type Terrain = 'JUNGLE' | 'MOUNTAIN' | 'SANDS' | 'WETLAND';
export type InvaderType = 'EXPLORER' | 'TOWN' | 'CITY';
export type Element = 'SUN' | 'MOON' | 'FIRE' | 'AIR' | 'WATER' | 'EARTH' | 'PLANT' | 'ANIMAL';
export type PowerSpeed = 'FAST' | 'SLOW';

export const INVADER_HEALTH: Record<InvaderType, number> = {
  EXPLORER: 1,
  TOWN: 2,
  CITY: 3,
};

export const INVADER_DAMAGE: Record<InvaderType, number> = {
  EXPLORER: 1,
  TOWN: 2,
  CITY: 3,
};

export const INVADER_FEAR: Record<InvaderType, number> = {
  EXPLORER: 0,
  TOWN: 1,
  CITY: 2,
};

export const DAHAN_HEALTH = 2;
export const DAHAN_DAMAGE = 2;

export interface Invader {
  type: InvaderType;
  damageTaken: number;
}

export function makeInvader(type: InvaderType): Invader {
  return { type, damageTaken: 0 };
}

export function invaderRemainingHealth(inv: Invader): number {
  return INVADER_HEALTH[inv.type] - inv.damageTaken;
}

export function invaderIsDestroyed(inv: Invader): boolean {
  return inv.damageTaken >= INVADER_HEALTH[inv.type];
}

export interface Dahan {
  damageTaken: number;
}

export function makeDahan(): Dahan {
  return { damageTaken: 0 };
}

export function dahanRemainingHealth(d: Dahan): number {
  return DAHAN_HEALTH - d.damageTaken;
}

export function dahanIsDestroyed(d: Dahan): boolean {
  return d.damageTaken >= DAHAN_HEALTH;
}

export interface PowerCard {
  name: string;
  cost: number;
  speed: PowerSpeed;
  range: number;
  target: string;
  elements: Element[];
  description: string;
}
